MergerLogic/DataTypes/Data.cs

```diff
@@ -198,25 +199,58 @@ protected virtual Extent GetExtent()
 
         protected virtual Tile? InternalGetLastExistingTile(Coord coords)
         {
-            this._logger.LogDebug($"[{MethodBase.GetCurrentMethod().Name}] started for coord: z:{coords.Z}, x:{coords.X}, y:{coords.Y}");
-            int z = coords.Z;
-            int baseTileX = coords.X;
-            int baseTileY = this.ConvertOriginCoord(coords); //dont forget to use the correct origin when overriding this
+            return Task.Run<Tile?>(() =>
+            {
+                this._logger.LogDebug($"[{MethodBase.GetCurrentMethod().Name}] started for coord: z:{coords.Z}, x:{coords.X}, y:{coords.Y}");
+                // get tiles coordinates
+                int z = coords.Z;
+                int baseTileX = coords.X;
+                int baseTileY = this.ConvertOriginCoord(coords); //dont forget to use the correct origin when overriding this
+
+                // Define all tiles coordinates that needs to be requested for upscale
+                List<Coord> coordsList = new List<Coord>(MaxZoomRead - (MaxZoomRead - coords.Z));
+                for (int i = z - 1; i >= 0; i--)
+                {
+                    baseTileX >>= 1; // Divide by 2
+                    baseTileY >>= 1; // Divide by 2
 
-            Tile? lastTile = null;
+                    coordsList.Add(new Coord(i, baseTileX, baseTileY));
+                }
+                var response = this.InternalGetExistingTile(coordsList.ToArray());
```

Comment:
Instead of adding this function please move it's code to here and remove it.

Reply:
fixed